Add ARM support

joshuanianji opened the pull request:
Closes #4 

This is a pretty quick fix to add arm support for versions above 0.2.0, although it doesn't check if the user has an architecture other than arm and amd. If you're fine with this, maybe we can add a short message to the NOTES.md going over supported architectures.

Maybe another way would be to check if `https://github.com/typst/typst/releases/download/v${TYPST_VERSION}/typst-$(uname -m)-unknown-linux-musl.tar.xz` is a valid url, and throwing an error if not. What do you think?

joshuanianji committed:
Add arm support

joshuanianji committed:
Update comment

joshuanianji committed:
Fix bash conditionals

michidk commented:
I actually prefer this offline check. Could we keep a list of every supported architecture and check against that if the version is larger than 0.2.0? In your implementation, we assume every architecture is valid as long as not 0.2.0 or 0.1.0.
We should also keep in mind, that they could potentially release a security fix under 0.2.1 for the old version in the future  (though unlikely).

joshuanianji commented:
Thanks for the update! I can finish up the offline architecture check.

Do you think a possible v0.2.1 won't add new architecture support? If we can assume so, I can just check if the string begins with "0.2" or "0.1".

joshuanianji committed:
Hardcode supported archs for v0.3.0 and beyond

joshuanianji committed:
Improve version check for 0.2 and 0.1

check if string starts with rather than checking the entire string

michidk commented:
> Do you think a possible v0.2.1 won't add new architecture support? If we can assume so, I can just check if the string begins with "0.2" or "0.1".

Yeah I would make that assumption

michidk commented on src/typst/install.sh:
```diff
@@ -93,14 +93,43 @@ find_version_from_git_tags() {
+    echo "(!) Architecture $architecture unsupported for Typst version v$TYPST_VERSION!"
     exit 1
+else
+    echo "Arch test finished. Typst should be good to go!"
```
```suggestion
    echo "Your platform architecture is supported by Typst version v$TYPST_VERSION!"
```

michidk commented:
Actually, below version 0.3.0 the script would not work anyways, because the linux binary was based on gnu/glibc and not musl (and therefore the filename different).
![image](https://github.com/michidk/devcontainers-features/assets/3979930/eb6c3716-ef10-47f8-814e-2c3a7b7f63f7)

But we (I) always assume that the filename contains 'musl":
https://github.com/michidk/devcontainers-features/blob/main/src/typst/install.sh#L112

Could you also implement that it will use the gnu version for versions below 0.3.0 on the amd64 platform? This would be great.

joshuanianji committed:
Update final `echo` statement after platform arch test

Co-authored-by: Michael Lohr <michael@lohr.dev>

joshuanianji commented:
> Could you also implement that it will use the gnu version for versions below 0.3.0 on the amd64 platform? This would be great.

Haha good catch, I can add that in. I also realized that versions 0.1 and 0.2 have `tar.gz` extensions, while 0.3 and beyond have `tar.xz`. I put both of these variations in a `LIBC_AND_EXTENSION` variable

joshuanianji committed:
Fix download link (libc and file extension)

This makes 0.1 and 0.2 work

michidk commented:
Thanks for the changes!
I get the following error with versions 0.1.0 and 0.2.0:
```bash
sudo VERSION=0.2.0 ./install.sh
TYPST_VERSION=0.2.0
Downloading typst version 0.2.0...
xz: (stdin): File format not recognized
tar: Child died with signal 13
tar: Error is not recoverable: exiting now
```

Since tar does not support compression detection when feeding from stdin (https://superuser.com/questions/689818/let-tar-auto-detect-compression-type-when-extracting-from-stdin), we have to have another if-branch.

For version 0.1.0 and 0.2.0 we have to use the `tar xfz` command and for all the other versions `tar xfJ`.

joshuanianji commented:
Thanks! I'll add that in.

I'll add tests for just `v0.2` and `v0.5`, since I couldn't find a way to easily matrix test custom scenarios. I'll also update the `proposals` section to have the minor versions from 0.1 to 0.5 as well as `latest`.

joshuanianji committed:
Add `v0.2` and `v0.5` version scenarios

joshuanianji committed:
Add all typst versions in the proposals

joshuanianji committed:
Use different tar commands for `gz` and `xz`

joshuanianji committed:
Split LIBC and EXTENSION variables

When we unpack the tar file, the folder contains the libc, but not the tar extension.
To make sure we reference the correct folder when we extract the binary file, we need to be able to have access to just the LIBC variable

joshuanianji committed:
Funky tests with `javascript-node` base img

joshuanianji commented:
I also found an issue with Typst versions 0.2 and 0.1 running on the devcontainer for this repo. 

The base image `mcr.microsoft.com/devcontainers/javascript-node:0-18` has a libc version of 2.31 (via `ldd --version`, and also [changelog](https://github.com/devcontainers/images/blob/main/src/javascript-node/history/0.205.2.md)), so although the install script works (I think), writing `typst --version` results in the following error:

```log
typst: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by typst)
typst: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by typst)
typst: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by typst)
typst: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by typst)
```

I'm pretty sure this is just a result of the feature starter template having an older version of the javascript-node image. When I used an updated version (either the 1-18 or 20 tag), everything worked fine. Since this might be a pretty confusing bug for others, do you want to put some sort of section in this feature's README? Maybe a "supported versions" section or something.

michidk commented:
Yeah that's an known issue (https://github.com/typst/typst/issues/109, fixed in https://github.com/typst/typst/pull/847) with Typst and won't be fixed for those versions.
I don't know exactly how this stuff works, but I would assume that on a system with the right glibc version, this typst executable would still work.

Great what you implemented there! Love the tests as well.

The `'javascript-node-0-18'` fails on the CI. I would remove this test. We maybe should really add a note to the README, referencing the issue and stating that those versions will only work with specific glib versions. Do you want to add this as well?

michidk committed:
Fix formatting

joshuanianji commented:
Thanks! I'll delete the `javascript-node-0-18` test and add a note

joshuanianji committed:
Delete `javascript-node-0-18` test (typst v0.2.0)

joshuanianji committed:
Add note on OS and Arch support

michidk commented:
Really cool! Thanks a lot for your changes!
